### PR TITLE
Output formats

### PR DESCRIPTION
**Pull request by qrilka**
This adds intermediate representation for formatted output. I added basic output limiting, also plan to add JSON output so it's WIP right now

---
This change is [Reviewable](https://reviewable.io/reviews/nh2/hatrace/22)

### CODE REVIEW

**Comment by qrilka** [comment]
Finally some JSON output - https://gist.github.com/qrilka/30780734c430bd90c51dadb836f36c9a

**Comment by qrilka** [comment]
Ready for a review, I will deal with "typed" JSON in a separate PR

File: src/System/Hatrace.hs [hunk 1]
```diff
@@ -1442,36 +1449,267 @@ atomicWritesSink =
...
+      putStrLn $ "Got signal: " ++ prettySignal sig
+
+    EventProcessDeath exitCode ->
+      putStrLn $ "Process exited with status: " ++ show exitCode
+
+
+printHatraceEventJson :: HatraceEvent -> IO ()
+printHatraceEventJson hatraceEvent = do
+  BS.putStr $ BSL.toStrict $ encode hatraceEvent
+  BS.putStr "\n"
```
Comment (nh2) [inline src/System/Hatrace.hs]: Probably want to do that in 1 `putStr` syscall to reduce chances of interleaving.

File: src/System/Hatrace/Format.hs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    FixedStringArg s -> toJSON s
+    VarLengthStringArg s -> toJSON s
+    ListArg xs -> toJSON xs
+    StructArg fieldValues ->
+      object [ T.pack name .= value | (name, value) <- fieldValues ]
+
+data FormattedReturn
+  = NoReturn
+  | FormattedReturn FormattedArg
+  deriving (Eq, Show)
```
Comment (nh2) [inline src/System/Hatrace/Format.hs]: Let's add `Ord` everywhere as well, so that such things can be put in sets and maps

**Comment by qrilka** [comment]
Updated new syscall guide and straightened history a bit (but unfortunately not every commit is buildable).
Ready for a merge.

**Comment by nh2** [comment]
> but unfortunately not every commit is buildable

@qrilka Then let's squash those that aren't buildable and don't necessarily need to be separate.

The motivation behind linear history is being able to bisect regressions automatically, and non-building commits prevent that.

**Comment by qrilka** [comment]
Ok, I squashed it a bit so now every commit builds fine

**Comment by nh2** [comment]
Great, thanks!

**Pull request merged**